StoreMessageContentActivity/handler.ts

```diff
@@ -135,62 +234,100 @@ export const getStoreMessageContentActivityHandler = (
     return { kind: "FAILURE", reason: "MASTER_INBOX_DISABLED" };
   }
 
-  // whether the user has blocked inbox storage for messages from this sender
-  const isMessageStorageBlockedForService =
-    blockedInboxOrChannels.indexOf(BlockedInboxOrChannelEnum.INBOX) >= 0;
+  //
+  // check Service Preferences Settings
+  //
+  return await getServicePreferenceValueOrError(lServicePreferencesModel)({
+    fiscalCode: newMessageWithoutContent.fiscalCode,
+    serviceId: newMessageWithoutContent.senderServiceId,
+    userServicePreferencesMode: profile.servicePreferencesSettings.mode,
+    userServicePreferencesVersion: profile.servicePreferencesSettings.version
+  })
+    .fold<ReadonlyArray<BlockedInboxOrChannelEnum>>(servicePreferenceError => {
+      if (servicePreferenceError.kind === "ERROR") {
+        // The query has failed, we consider this as a transient error.
+        context.log.error(`${logPrefix}|${servicePreferenceError.message}`);
+        throw Error("Error while retrieving user's service preference");
+      }
 
-  if (isMessageStorageBlockedForService) {
-    // the recipient's inbox is disabled
-    context.log.warn(`${logPrefix}|RESULT=SENDER_BLOCKED`);
-    return { kind: "FAILURE", reason: "SENDER_BLOCKED" };
-  }
+      // channels the user has blocked for this sender service
+      const blockedInboxOrChannels = fromNullable(
+        profile.blockedInboxOrChannels
+      )
+        .chain(bc => fromNullable(bc[newMessageWithoutContent.senderServiceId]))
+        .getOrElse([]);
 
-  // Save the content of the message to the blob storage.
-  // In case of a retry this operation will overwrite the message content with itself
-  // (this is fine as we don't know if the operation succeeded at first)
-  const errorOrAttachment = await lMessageModel
-    .storeContentAsBlob(
-      lBlobService,
-      newMessageWithoutContent.id,
-      createdMessageEvent.content
-    )
-    .run();
+      context.log.verbose(
+        `${logPrefix}|BLOCKED_CHANNELS=${JSON.stringify(
+          blockedInboxOrChannels
+        )}`
+      );
 
-  if (isLeft(errorOrAttachment)) {
-    context.log.error(`${logPrefix}|ERROR=${errorOrAttachment.value}`);
-    throw new Error("Error while storing message content");
-  }
+      // an error occurs also when user service preference mode is LEGACY
+      context.log.warn(`${logPrefix}|${servicePreferenceError.message}`);
 
-  // Now that the message content has been stored, we can make the message
-  // visible to getMessages by changing the pending flag to false
-  const updatedMessageOrError = await lMessageModel
-    .upsert({
-      ...newMessageWithoutContent,
-      isPending: false
-    })
-    .run();
+      return blockedInboxOrChannels;
+    }, identity)
+    .map<Promise<StoreMessageContentActivityResult>>(
```

Comment:
Why not chain this with subsequent tasks in a more functional way. I'm wondering about manage Promise using `await` inside map operation, while having the possibility to use `chain` capability. 
AFAIK, in alternative we can mantain a more procedural style without using `chain` construct as  the previous version :). What do you think?

Reply:
To comply with new requests I decided to keep the old "good" code as a "block" to use where it was necessary, I can take a look to transform the old code to a new and more functional one. I'll let you know.

Reply:
> I decided to keep the old "good" code as a "block"

I think this is a good design decision. 

> a new and more functional one

That would require a huge refactor, are you sure it is worth the effort?

I think we can have best of both world by doing the following:
* write new code using FP, and isolate that in a well-scoped function (you did it already in `getServicePreferenceValueOrError`)
* use such function inside the _legacy_ imperative code

Something like:
```ts
const remappedBlockedInboxOrChannels = 
    await getServicePreferenceValueOrError(/* whatever */)
		.fold(/* in case you need it */)
		.run();

// whether the user has blocked inbox storage for messages from this sender
const isMessageStorageBlockedForService =
	remappedBlockedInboxOrChannels.indexOf(
		BlockedInboxOrChannelEnum.INBOX
	) >= 0;
	
// ...and code follows from here...
```	
	
	
@AleDore @michaeldisaro what do you think?

Reply:
I'm just trying to do that refactor, if I'll close it for this afternoon without test errors it could be a good chance to have better functional code. At least I tried...

Reply:
@balanza I made some refactoring, not as deep as @AleDore suggested, in https://github.com/pagopa/io-functions-services/pull/142/commits/b232fc7bbe7952e787c06e39f0450557e671ae0e and previous commits.